### PR TITLE
Feature: Chat History export to JSON

### PR DESCRIPTION
**Pull request by alashchev17**
# Feature: Chat History export to JSON

## Description

- Problem: Chat history export is not implemented yet.
- Solution: Chat history export of a specific chat thread is implemented.
- [Any background context or related links?](https://github.com/orgs/smallcloudai/projects/5/views/1?pane=issue&itemId=77086911)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, only code improvements)
- [ ] Documentation update

## How to Test

- Step 1: Create a New Chat
- Step 2: Ask a chat model something
- Step 3: Click a tiny button near the paper plane send button within a chat form
- Step 4: Inspect the JSON that you got on a new page
### Thread history button
Thread history button
### Thread history page
Thread history page

## Checklist

- [x] My code follows the code style of this project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I have updated the documentation where necessary.

### CODE REVIEW

**Comment by alashchev17** [comment]
Work is still in progress

File: src/features/ThreadHistory/ThreadHistory.module.css [hunk 1]
```diff
@@ -0,0 +1,10 @@
+.calloutContainer {
+  position: absolute;
+  left: 50%;
+  top: 30px;
+  transform: translateX(-50%);
+  width: fit-content;
+  z-index: 1;
+  display: flex;
+  justify-content: center;
+}
```
Comment (MarcMcIntosh) [inline src/features/ThreadHistory/ThreadHistory.module.css:10]: is the pre selector used here?

**Review by MarcMcIntosh — COMMENTED** [review]
A couple questions and a few things to clean up :)

File: src/lib/render/web.css [hunk 2]
```diff
@@ -7,3 +7,7 @@ body {
   width: 260px;
   flex-shrink: 0;
 }
+
+pre {
```
[the remark this answers is not in the document]
Comment (alashchev17) [reply src/lib/render/web.css:11]: This one is to reset margins for <pre> elements inside of codeblocks, to make the actual box look better

File: src/features/ThreadHistory/ThreadHistory.module.css (hunk 1 above)
Comment (alashchev17) [reply src/features/ThreadHistory/ThreadHistory.module.css:10]: No, pre selector is used in a code block area, `calloutContainer` is used only for info or error bubbles

File: src/features/App.tsx [hunk 3]
```diff
@@ -187,6 +188,15 @@ export const InnerApp: React.FC<AppProps> = ({ style }: AppProps) => {
             onCloseStatistic={goBack}
           />
         )}
+        {page.name === "thread history page" && (
+          <ThreadHistory
+            backFromThreadHistory={goBack}
+            tabbed={config.tabbed}
+            host={config.host}
+            onCloseThreadHistory={goBack}
+            chatId={page.chatId}
```
[the remark this answers is not in the document]
Comment (alashchev17) [reply src/features/App.tsx:197]: Yes, it is

File: src/components/Toolbar/Dropdown.tsx [hunk 4]
```diff
@@ -20,6 +20,7 @@ export type DropdownNavigationOptions =
   | "hot keys"
   | "restart tour"
   | "cloud login"
+  | "chat history"
```
[the remark this answers is not in the document]
Comment (alashchev17) [reply src/components/Toolbar/Dropdown.tsx:23]: No, I'll remove it now

File: src/components/Chat/Chat.tsx [hunk 5]
```diff
@@ -112,6 +139,13 @@ export const Chat: React.FC<ChatProps> = ({
         ref={chatContentRef}
         onRetry={retryFromIndex}
       />
+      {/* {isViewingRawJSON && (
```
[the remark this answers is not in the document]
Comment (alashchev17) [reply src/components/Chat/Chat.tsx:142]: Yeah, sure

**Comment by alashchev17** [comment]
Removed unused and commented code, added a new state to manage visibility of this button (should click on mode first, then button appears)

File: src/lib/render/web.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarcMcIntosh) [reply src/lib/render/web.css:11]: can we scope it to the code block area?

[the remark this answers is not in the document]
Comment (MarcMcIntosh) [reply src/lib/render/web.css:11]: I'll also need to check this once it's merged because the 📍 messages also use `pre`

**Review by MarcMcIntosh — COMMENTED** [review]
Scope the `pre` selector so it's not global, and merge the alpha branch in because the 📍 PR also uses the `pre` tag with some css applied :)

**Review by MarcMcIntosh — APPROVED** [review]
LGTM :)

**Pull request merged**